### ISSUE TITLE
Client script runs multiple times without __main__

### ISSUE THREAD

**Issue by thesamovar** [workhorsy/py-cpuinfo#108]
The following script reproduces the problem:

```
import cpuinfo
cpuinfo.get_cpu_info()
print 'How many times should I see this?'
```

gives output

```
How many times should I see this?
How many times should I see this?
```

I believe the problem was introduced in https://github.com/workhorsy/py-cpuinfo/commit/9bbd7fb8adb30c1cfb4389dd696a758d9d1f172f and is because the function ``_get_cpu_info_from_cpuid`` uses multiprocessing (which is problematic on Windows).

Environment is Windows 7, 64 bit, using a freshly downloaded Anaconda 2.7. Full package environment below:

```
# packages in environment at C:\Anaconda:
#
# Name                    Version                   Build  Channel
_ipyw_jlab_nb_ext_conf    0.1.0                    py27_0  
alabaster                 0.7.11                   py27_0  
anaconda                  custom           py27h689e5c3_0  
anaconda-client           1.7.1                    py27_0  
anaconda-navigator        1.8.7                    py27_0  
anaconda-project          0.8.2                    py27_0  
appdirs                   1.4.3            py27h28b3542_0  
asn1crypto                0.24.0                   py27_0  
astroid                   1.6.5                    py27_0  
astropy                   2.0.8            py27hc997a72_1  
atomicwrites              1.1.5                    py27_0  
attrs                     18.1.0                   py27_0  
automat                   0.7.0                    py27_0  
babel                     2.6.0                    py27_0  
backports                 1.0                      py27_1  
backports.functools_lru_cache 1.5                      py27_1  
backports.shutil_get_terminal_size 1.0.0                    py27_2  
backports.shutil_which    3.5.1                    py27_2  
backports_abc             0.5              py27h0ec6b72_0  
beautifulsoup4            4.6.3                    py27_0  
bitarray                  0.8.3            py27h0c8e037_0  
bkcharts                  0.2              py27h92b6de3_0  
blas                      1.0                         mkl  
blaze                     0.11.3                   py27_0  
bleach                    2.1.3                    py27_0  
blosc                     1.14.4               h95ae18b_0  
bokeh                     0.13.0                   py27_0  
boto                      2.49.0                   py27_0  
bottleneck                1.2.1            py27hc997a72_1  
bzip2                     1.0.6                h0c8e037_5  
ca-certificates           2018.03.07                    0  
cdecimal                  2.3              py27h0c8e037_3  
certifi                   2018.8.13                py27_0  
cffi                      1.11.5           py27hdb016f4_0  
chardet                   3.0.4                    py27_1  
click                     6.7              py27hb6defca_0  
cloudpickle               0.5.3                    py27_0  
clyent                    1.2.2                    py27_1  
colorama                  0.3.9            py27hdfe4ae1_0  
comtypes                  1.1.7                    py27_0  
conda                     4.5.10                   py27_0  
conda-build               3.12.1                   py27_0  
conda-env                 2.6.0                         1  
conda-verify              3.1.0                    py27_0  
configparser              3.5.0            py27h2fa79a8_0  
console_shortcut          0.1.1                         3  
constantly                15.1.0           py27h28b3542_0  
contextlib2               0.5.5            py27h42efda5_0  
cryptography              2.3              py27hc64555f_0  
cryptography-vectors      2.3                      py27_0  
curl                      7.61.0               h7a46e7a_0  
cycler                    0.10.0           py27h59acbbf_0  
cython                    0.28.5           py27hc56fc5f_0  
cytoolz                   0.9.0.1          py27h0c8e037_1  
dask                      0.18.2                   py27_0  
dask-core                 0.18.2                   py27_0  
datashape                 0.5.4                    py27_1  
decorator                 4.3.0                    py27_0  
distributed               1.22.1                   py27_0  
docutils                  0.14             py27h8652d09_0  
entrypoints               0.2.3                    py27_2  
enum34                    1.1.6                    py27_1  
et_xmlfile                1.0.1            py27h1de5d23_0  
fastcache                 1.0.2            py27h0c8e037_2  
filelock                  3.0.4                    py27_0  
flask                     1.0.2                    py27_1  
flask-cors                3.0.6                    py27_0  
freetype                  2.9.1                h4d385ea_1  
funcsigs                  1.0.2            py27h8885ae1_0  
functools32               3.2.3.2                  py27_1  
future                    0.16.0                   py27_1  
futures                   3.2.0                    py27_0  
get_terminal_size         1.0.0                h38e98db_0  
gevent                    1.3.5            py27h0c8e037_0  
glob2                     0.6                      py27_0  
greenlet                  0.4.14           py27h0c8e037_0  
grin                      1.2.1                    py27_4  
gsl                       2.4                  h0c8e037_4  
h5py                      2.8.0            py27h5f1a774_1  
hdf5                      1.10.2               h4b43f98_1  
heapdict                  1.0.0                    py27_2  
html5lib                  1.0.1                    py27_0  
hyperlink                 18.0.0                   py27_0  
icc_rt                    2017.0.4             h97af966_0  
icu                       58.2                 h2aa20d9_1  
idna                      2.7                      py27_0  
imageio                   2.3.0                    py27_0  
imagesize                 1.0.0                    py27_0  
incremental               17.5.0                   py27_0  
intel-openmp              2018.0.3                      0  
ipaddress                 1.0.22                   py27_0  
ipykernel                 4.8.2                    py27_0  
ipython                   5.8.0                    py27_0  
ipython_genutils          0.2.0            py27hbe997df_0  
ipywidgets                7.4.0                    py27_0  
isort                     4.3.4                    py27_0  
itsdangerous              0.24                     py27_1  
jdcal                     1.4                      py27_0  
jedi                      0.12.1                   py27_0  
jinja2                    2.10                     py27_0  
jpeg                      9b                   ha175dff_2  
jsonschema                2.6.0            py27haaf3834_0  
jupyter                   1.0.0                    py27_5  
jupyter_client            5.2.3                    py27_0  
jupyter_console           5.2.0                    py27_1  
jupyter_core              4.4.0                    py27_0  
jupyterlab                0.33.11                  py27_0  
jupyterlab_launcher       0.11.2                   py27_0  
keyring                   13.2.1                   py27_0  
kiwisolver                1.0.1            py27hc56fc5f_0  
lazy-object-proxy         1.3.1            py27h0c8e037_2  
libcurl                   7.61.0               h7a46e7a_0  
libiconv                  1.15                 hda2e4ec_7  
libpng                    1.6.34               h325896a_0  
libsodium                 1.0.16               h8b3e59e_0  
libssh2                   1.8.0                h77a7533_4  
libtiff                   4.0.9                h1c3b264_1  
libxml2                   2.9.8                hdfd6920_1  
libxslt                   1.1.32               h89dfad8_0  
linecache2                1.0.0                    py27_0  
llvmlite                  0.24.0           py27hc56fc5f_0  
locket                    0.2.0            py27h1ca288a_1  
lxml                      4.2.4            py27h31b8cb8_0  
lzo                       2.10                 h0bb7fe3_2  
m2w64-gcc-libgfortran     5.3.0                         6  
m2w64-gcc-libs            5.3.0                         7  
m2w64-gcc-libs-core       5.3.0                         7  
m2w64-gmp                 6.1.0                         2  
m2w64-libwinpthread-git   5.0.0.4634.697f757               2  
markupsafe                1.0              py27h0c8e037_1  
matplotlib                2.2.3            py27h263d877_0  
mccabe                    0.6.1                    py27_1  
menuinst                  1.4.14           py27h0c8e037_0  
mistune                   0.8.3            py27h0c8e037_1  
mkl                       2018.0.3                      1  
mkl-service               1.1.2            py27h3f2f999_4  
mkl_fft                   1.0.4            py27h31cbb53_1  
more-itertools            4.3.0                    py27_0  
mpmath                    1.0.0                    py27_2  
msgpack-python            0.5.6            py27hdc96acc_1  
msys2-conda-epoch         20160418                      1  
multipledispatch          0.6.0                    py27_0  
navigator-updater         0.2.1                    py27_0  
nbconvert                 5.3.1                    py27_0  
nbformat                  4.4.0            py27hf49b375_0  
networkx                  2.1                      py27_0  
nltk                      3.3.0                    py27_0  
nose                      1.3.7                    py27_2  
notebook                  5.6.0                    py27_0  
numba                     0.39.0           py27h39f3610_0  
numexpr                   2.6.7            py27h8d23284_0  
numpy                     1.15.0           py27h911edcf_0  
numpy-base                1.15.0           py27hfef472a_0  
numpydoc                  0.8.0                    py27_0  
odo                       0.5.1            py27h64810b2_0  
olefile                   0.45.1                   py27_0  
openpyxl                  2.5.5                    py27_0  
openssl                   1.0.2p               h0c8e037_0  
packaging                 17.1                     py27_0  
pandas                    0.23.4           py27h39f3610_0  
pandoc                    2.2.1                h1a437c5_0  
pandocfilters             1.4.2                    py27_1  
parso                     0.3.1                    py27_0  
partd                     0.3.8            py27h1e0692f_0  
path.py                   11.0.1                   py27_0  
pathlib2                  2.3.2                    py27_0  
patsy                     0.5.0                    py27_0  
pep8                      1.7.1                    py27_0  
pickleshare               0.7.4            py27hb5f6335_0  
pillow                    5.2.0            py27h5b88493_0  
pip                       10.0.1                   py27_0  
pkginfo                   1.4.2                    py27_1  
pluggy                    0.7.1                    py27_0  
ply                       3.11                     py27_0  
prometheus_client         0.3.1                    py27_0  
prompt_toolkit            1.0.15           py27h3a8ec6a_0  
psutil                    5.4.6            py27h0c8e037_0  
py                        1.5.4                    py27_0  
py-cpuinfo                4.0.0                    py27_0  
pyasn1                    0.4.4                    py27_0  
pyasn1-modules            0.2.2                    py27_0  
pycodestyle               2.4.0                    py27_0  
pycosat                   0.6.3            py27h0c8e037_0  
pycparser                 2.18                     py27_1  
pycrypto                  2.6.1            py27h0c8e037_9  
pycurl                    7.43.0.2         py27hc64555f_0  
pyflakes                  2.0.0                    py27_0  
pygments                  2.2.0            py27ha50f84f_0  
pylint                    1.9.2                    py27_0  
pyodbc                    4.0.23           py27hc56fc5f_0  
pyopenssl                 18.0.0                   py27_0  
pyparsing                 2.2.0                    py27_1  
pyqt                      5.6.0            py27h6e61f57_6  
pysocks                   1.6.8                    py27_0  
pytables                  3.4.4            py27hdd5b532_0  
pytest                    3.7.1                    py27_0  
python                    2.7.15               he216670_0  
python-dateutil           2.7.3                    py27_0  
pytz                      2018.5                   py27_0  
pywavelets                0.5.2            py27hc997a72_2  
pywin32                   223              py27h0c8e037_1  
pywinpty                  0.5.4                    py27_0  
pyyaml                    3.13             py27h0c8e037_0  
pyzmq                     17.1.2           py27h0c8e037_0  
qt                        5.6.2            vc9hc26998b_12  [vc9]
qtawesome                 0.4.4            py27h766b13d_0  
qtconsole                 4.4.0                    py27_0  
qtpy                      1.4.2                    py27_1  
requests                  2.19.1                   py27_0  
rope                      0.11.0                   py27_0  
ruamel_yaml               0.15.46          py27h0c8e037_0  
scandir                   1.9.0            py27h0c8e037_0  
scikit-image              0.14.0           py27hc56fc5f_1  
scikit-learn              0.19.1           py27hbe35404_0  
scipy                     1.1.0            py27ha6b90e7_0  
seaborn                   0.9.0                    py27_0  
send2trash                1.5.0                    py27_0  
service_identity          17.0.0           py27h28b3542_0  
setuptools                40.0.0                   py27_0  
simplegeneric             0.8.1                    py27_2  
singledispatch            3.4.0.3          py27h3f9d112_0  
sip                       4.18.1           py27hc56fc5f_2  
six                       1.11.0                   py27_1  
snappy                    1.1.7                he46498f_3  
snowballstemmer           1.2.1            py27h28d3bf7_0  
sortedcollections         1.0.1                    py27_0  
sortedcontainers          2.0.4                    py27_0  
sphinx                    1.7.6                    py27_0  
sphinxcontrib             1.0                      py27_1  
sphinxcontrib-websupport  1.1.0                    py27_1  
spyder                    3.3.1                    py27_1  
spyder-kernels            0.2.6                    py27_0  
sqlalchemy                1.2.10           py27h0c8e037_0  
sqlite                    3.24.0               h7a46e7a_0  
ssl_match_hostname        3.5.0.1                  py27_2  
statsmodels               0.9.0            py27hc997a72_0  
subprocess32              3.5.2            py27h0c8e037_0  
sympy                     1.2                      py27_0  
tblib                     1.3.2            py27h8ae915c_0  
terminado                 0.8.1                    py27_1  
testpath                  0.3.1            py27h1cd488d_0  
tk                        8.6.7                h144d9c4_3  
toolz                     0.9.0                    py27_0  
tornado                   5.1              py27h0c8e037_0  
traceback2                1.4.0                    py27_0  
traitlets                 4.3.2            py27h1b1b3a5_0  
twisted                   18.7.0           py27h0c8e037_1  
typing                    3.6.4                    py27_0  
unicodecsv                0.14.1           py27h0bf7bb0_0  
unittest2                 1.1.0                    py27_0  
urllib3                   1.23                     py27_0  
vc                        9                    h7299396_1  
vs2008_runtime            9.00.30729.1         hfaea7d5_1  
vs2015_runtime            15.5.2                        3  
wcwidth                   0.1.7            py27hb1a0d82_0  
weave                     0.16.0                   py27_0    conda-forge
webencodings              0.5.1                    py27_1  
werkzeug                  0.14.1                   py27_0  
wheel                     0.31.1                   py27_0  
widgetsnbextension        3.4.0                    py27_0  
win_inet_pton             1.0.1                    py27_1  
win_unicode_console       0.5              py27hc037021_0  
wincertstore              0.2              py27hf04cefb_0  
winpty                    0.4.3                         4  
wrapt                     1.10.11          py27h0c8e037_2  
xlrd                      1.1.0                    py27_1  
xlsxwriter                1.0.5                    py27_0  
xlwings                   0.11.8                   py27_0  
xlwt                      1.3.0            py27h2271735_0  
yaml                      0.1.7                h3e6d941_2  
zeromq                    4.2.5                h2880e7c_1  
zict                      0.1.3                    py27_0  
zlib                      1.2.11               hbc2faf4_2  
zope                      1.0                      py27_0  
zope.interface            4.5.0            py27h0c8e037_0  
```

**Comment by workhorsy** [comment]
Interesting. I was able to reproduce on Windows 10 with Python 2.7.15 and  3.7.0. I'll see if I can figure out what is wrong. Thanks.

**Comment by thesamovar** [comment]
I think the issue is that when you use multiprocessing on Windows, because Windows doesn't support forking, it fires up a new process and so the script gets re-run: https://docs.python.org/2/library/multiprocessing.html#windows

**Comment by workhorsy** [comment]
I finally remembered why this is running the script twice. Basically Python multiprocessing is the only reliable way to run code safely in another process that may segfault. But it has the design flaw that any script will be executed multiple times. So the only solution I came up with, was to require the use of a ```__main__```.

```python
if __name__ == '__main__':
    import cpuinfo
    cpuinfo.get_cpu_info()
    print 'How many times should I see this?'
```

Obviously this is not an ideal solution, but it is the best I can come up with given the limitations of forking Python on Windows. Does this solve your problem?

**Comment by thesamovar** [comment]
Thanks for investigating. That's not a possible solution for us, because it would mean requiring all of our Windows users to change their existing code. However, version 3.3 still works fine for our needs, so I'm wondering if there could be a flag added so that whatever features added in 4.0 that required this change could be ignored? That would be much better for us than requiring ``<=3.3`` in our setup file.

**Comment by thesamovar** [comment]
Would it be possible to add something to ``py-cpuinfo`` so that it does everything except for the bit that's done in multiprocessing so that we don't have to require ``<=3.3``?

**Comment by thesamovar** [comment]
Follow-up, since we're still having difficulties with this. If the only thing I want is ``cpuinfo.get_cpu_info()['flags']`` is it safe to do the ``cpuinfo.DataSource.can_cpuid = False`` trick? And will that trick work with all versions of ``py-cpuinfo`` or only recent ones?

**Comment by workhorsy** [comment]
The `cpuinfo.DataSource.can_cpuid = False` should work on py-cpuinfo 3.0.0 and newer.

**Comment by thesamovar** [comment]
And removing ``can_cpuid`` won't impact the accuracy of flags?

**Comment by workhorsy** [comment]
It may show less flags on some systems.

**Comment by workhorsy** [comment]
For example, on a Windows machine with no wmic, it would only have access to the registry. And the registry can see CPU flags like SSE2, but not SSE3.

**Comment by thesamovar** [comment]
Hmm, OK it's complicated then. The flags I'm interested in specifically are SSE, SSE2, AVX and AVX2 because I want to set MSVC's ``/arch`` flag to the highest possible. Would those all work on all systems? If not, which systems wouldn't they work on?

**Comment by workhorsy** [comment]
The problem is that there is no guarantee that any flags will be returned. I don't think there is way to get newer cpu flags like AVX and AVX2 without CPUID.

**Comment by workhorsy** [comment]
How are you running py-cpuinfo? Is it possible for you to run it in a subscript? like by just running `python cpuinfo/cpuinfo.py` as a sub process? That would avoid the need to disable CPUID and work withput __main__

**Comment by thesamovar** [comment]
That sounds like a reasonable workaround (and we can even cache the result so it only has to be run once per computer). @mstimberg

**Comment by mstimberg** [comment]
@workhorsy, thanks for the advice and suggestion. I think we'll do something similar but simpler (as suggested by @thesamovar a while ago, if I remember correctly): we run a trivial custom-written script that has the `__main__` block:
```
import json

if __name__ == '__main__':
    import cpuinfo
    flags = cpuinfo.get_cpu_info()['flags']
    print(json.dumps(flags))
```
We can then call this script with something like `subprocess.check_output([sys.exectuable, 'our_script.py'])` and get a JSON string back. This feels too be safer than relying on the details of cpuinfo's script output.

**Comment by workhorsy** [comment]
Nice. I hope this works out for you. I'll have to see about adding a json option for running the script directly.

**Comment by workhorsy** [comment]
I'm working on a potential way to fix this. If I have more time, I may be able to fix this soon. Sorry that I can't give you a more accurate timeline. I've been extremely busy lately.

**Comment by workhorsy** [comment]
This should be fixed now. I'm changing things so a main is only required if using PyInstaller. So if using PyInstaller you wold need to do:

~~~python
# Required to use PyInstaller
if __name__ == '__main__':
	from multiprocessing import freeze_support
	from cpuinfo import get_cpu_info

	freeze_support()
	print(get_cpu_info())
~~~

Otherwise you can just do:

~~~python
from cpuinfo import get_cpu_info
print(get_cpu_info())
~~~

**Comment by workhorsy** [comment]
Fixed in #116

**Comment by thesamovar** [comment]
Great news! Thanks for your work on this.
